lib/client/conntest/ssh.go

```diff
@@ -49,30 +60,212 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
 		return nil, trace.Wrap(err)
 	}
 
-	// TODO(marco): test ssh connection
-	// - create certificate for user with extra field
-	// - create ssh client using that certificate
-	// - if connection fails because of a network error, a trace must be included indicating that the host is not reachable
-	// - other traces will be added by the Node itself (rbac checks, principal)
-	connectionDiagnostic.SetMessage("dry-run")
-	connectionDiagnostic.SetSuccess(true)
+	sshNode, err := s.clt.GetNode(ctx, defaults.Namespace, req.ResourceName)
+	if err != nil {
+		connDiag, err := s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			uuid.NewString(),
+			types.DiagnosticTraceTypeRBACNode,
+			"Node not found. Ensure the Node exists and your role allows you to access it.",
+			err,
+		))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	_, err = s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewSuccessTraceConnectionDiagnostic(
+		uuid.NewString(),
+		types.DiagnosticTraceTypeRBACNode,
+		"Resource exists.",
+	))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sshNodeAddr := sshNode.GetAddr()
+
+	key, err := client.NewKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	currentUser, err := s.clt.GetCurrentUser(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	certs, err := s.clt.GenerateUserCerts(ctx, proto.UserCertsRequest{
+		NodeName:               sshNode.GetName(),
+		PublicKey:              key.Pub,
+		Username:               currentUser.GetName(),
+		Expires:                time.Now().Add(time.Minute).UTC(),
+		Format:                 constants.CertificateFormatStandard,
+		Usage:                  proto.UserCertsRequest_SSH,
+		ConnectionDiagnosticID: connectionDiagnosticID,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	key.Cert = certs.SSH
+
+	// get certificate authorities
+	certAuths, err := s.clt.GetCertAuthorities(ctx, types.HostCA, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var certPublicKeys []ssh.PublicKey
+	for _, ca := range certAuths {
+		caCheckers, err := libsshutils.GetCheckers(ca)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		certPublicKeys = append(certPublicKeys, caCheckers...)
+	}
+
+	hostKeyCallback, err := sshutils.NewHostKeyCallback(sshutils.HostKeyCallbackConfig{
+		GetHostCheckers: func() ([]ssh.PublicKey, error) {
+			return certPublicKeys, nil
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sshClientConfig, err := key.ProxyClientSSHConfig(nil, sshNodeAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sshClientConfig.HostKeyCallback = hostKeyCallback
+	sshClientConfig.User = req.SSHPrincipal
+
+	dialCtx, cancelFunc := context.WithTimeout(ctx, req.DialTimeout)
+	defer cancelFunc()
+
+	var dialer net.Dialer
+	conn, err := dialer.DialContext(dialCtx, "tcp", sshNodeAddr)
```

Comment:
Hmm, are we connecting to the node directly? We should be connecting through the proxy, like `tsh ssh` does. Node may not be available directly e.g. if it's connected to the cluster via reverse tunnel.

Can we reuse TeleportClient/ProxyClient that we have in `lib/client`?